src/reflect/scala/reflect/internal/Definitions.scala

```diff
@@ -813,16 +843,27 @@ trait Definitions extends api.StandardDefinitions {
      * The class defining the method is a supertype of `tp` that
      * has a public no-arg primary constructor.
      */
-    def samOf(tp: Type): Symbol = if (!settings.Xexperimental) NoSymbol else findSam(tp)
-
-    def findSam(tp: Type): Symbol = {
-      // if tp has a constructor, it must be public and must not take any arguments
-      // (not even an implicit argument list -- to keep it simple for now)
-      val tpSym  = tp.typeSymbol
-      val ctor   = tpSym.primaryConstructor
-      val ctorOk = !ctor.exists || (!ctor.isOverloaded && ctor.isPublic && ctor.info.params.isEmpty && ctor.info.paramSectionCount <= 1)
+    def samOf(tp: Type): Symbol = {
+      // look at erased type because we (only) care about what ends up in bytecode
+      // (e.g., an alias type or intersection type is fine as long as the intersection dominator compiles to an interface)
+      val tpSym = erasure.javaErasure(tp).typeSymbol
+
+      if (tpSym.exists
+          // We use Java's MetaLambdaFactory, which requires an interface for the sam's owner
+          // (TODO: Can't use isInterface, yet, as it hasn't been updated for the new trait encoding)
+          && (tpSym.isJavaInterface || tpSym.isTrait)
```

Comment:
I've been back and forth on this -- happy to flip back the other way :-) How about making this platform specific and being more liberal when targeting JS?